src/SignalR/clients/csharp/Client.SourceGenerator/src/HubProxyGenerator.cs

```diff
@@ -0,0 +1,247 @@
+                                "SignalR.Client.SourceGenerator",
+                                DiagnosticSeverity.Warning,
+                                true), memberAccess.GetLocation(), hubSymbol.Name, member.Name));
+                        // Location for the diagnostic report is the GetProxy call site since there is no guarantee that
+                        //  the method definition lays within source of current project.
+                    }
+                    else
+                    {
+                        call = $"return this.conn.{specificCall}(\"{member.Name}\"{callArgs.ToString()});";
+                    }
+
+                    proxiesBody.AppendLine($@"
+        {signature}
+        {{
+            {call}
+        }}");
+                }
+
+                proxiesBody.AppendLine(@"
+    }");
+            }
+
+            proxiesBody.AppendLine(@"
+}");
+
+            // Populate GetProxy<T> with calls to appropriate proxy constructors
+            foreach (var (hubSymbol, memberAccess) in hubSymbols.Values)
+            {
+                var typeName = $"Microsoft.AspNetCore.SignalR.Client.SourceGenerated.Generated{hubSymbol.Name}";
+                getProxyBody.AppendLine($@"
+            if(typeof(THub) == typeof({hubSymbol})) return (THub) ({hubSymbol}) new {typeName}(conn);");
+            }
+
+            getProxyBody.AppendLine(getProxyFooter);
+            getProxyBody.AppendLine(proxiesBody.ToString());
+
+            context.AddSource("GeneratedBy" + nameof(HubProxyGenerator),
+                SourceText.From(getProxyBody.ToString(), Encoding.UTF8));
+        }
+
+        private string? getSpecificCall(IMethodSymbol member)
+        {
+            if (member.ReturnType is INamedTypeSymbol { Arity: 1, Name: "Task"} a
+                && a.TypeArguments[0] is INamedTypeSymbol { Arity: 1, Name: "ChannelReader"} b)
+            {
+                return $"StreamAsChannelAsync<{b.TypeArguments[0]}>";
+            }
+
+            if (member.ReturnType is INamedTypeSymbol { Arity: 1, Name: "IAsyncEnumerable"} c)
+            {
+                return $"StreamAsync<{c.TypeArguments[0]}>";
+            }
+
+            if (!member.ReturnType.ToString().StartsWith("System.Threading.Tasks.Task", StringComparison.Ordinal))
+            {
+                return null;
+            }
+
+            if (member.Parameters.Any(symbol =>
+                symbol.Type.ToString().Contains("ChannelReader<")
+                || symbol.Type.ToString().Contains("IAsyncEnumerable<")))
+            {
+                return "SendAsync";
+            }
+
+            if (member.ReturnType is INamedTypeSymbol {Arity: 1, Name: "Task"} d)
+            {
+                return $"InvokeAsync<{d.TypeArguments[0]}>";
+            }
+
+            if (member.ReturnType is INamedTypeSymbol {Arity: 0, Name: "Task"})
+            {
+                return "InvokeAsync";
```

Comment:
> there isn't an obvious way to distinguish between people wanting to use InvokeAsync vs. SendAsync.

Not for a return type of `Task` yeah... 

> Maybe we assume SendAsync for Task methods and of course InvokeAsync for value returning methods.

With `SendAsync`, the user is stuck with that behavior. At least with `InvokeAsync`, the user can ignore the task if they don't want to await roundtrip + processing by hub. So I'd prefer we generate `InvokeAsync` where we can't tell otherwise (and yes, at this point there is no way to tell :/ )

> Additionally, ValueTask/ValueTask<T> is supported for Hub methods, so we probably want to support that with the generator

Yes, we should definitely support. The implementation would be a `ValueTask` wrapping a `Task` since the `HubConnection` methods only return `Task`. I don't think `HubConnection` needs native `ValueTask` returns since SignalR calls are always to network resources.